crates/scheduler/src/bin/hypha-scheduler.rs

```diff
@@ -281,128 +260,202 @@ async fn run(config: ConfigWithMetadata<Config>) -> Result<()> {
-            let worker_task = Task::try_new(
-                network.clone(),
-                JobSpec {
-                    job_id,
-                    executor: TrainExecutorDescriptor::new(TRAIN_EXECUTOR_NAME)
-                        .into_executor(TrainExecutorConfig {
-                            model: diloco_config.model.clone().into(),
-                            data: Fetch::scheduler(peer_id, dataset.to_string()),
-                            updates: Send::peers(
-                                vec![parameter_server.peer_id()],
-                                SelectionStrategy::All,
-                            ),
-                            results: Receive::peers(vec![parameter_server.peer_id()]),
-                            optimizer: diloco_config.inner_optimizer.clone(),
-                            batch_size,
-                            preprocessor: diloco_config.preprocessor.clone().map(|p| p.into()),
-                            scheduler: None,
-                        })
-                        .into(),
-                },
-                &[w],
-            )
-            .await
-            .into_diagnostic()?;
-            worker_tasks.push(worker_task);
+        match max_batch_size_cfg {
+            Some(max) => raw.min(max),
+            None => raw,
         }
+    });
 
-        let _parameter_server_task = Task::try_new(
-            network.clone(),
-            JobSpec {
-                job_id,
-                executor: AggregateExecutorDescriptor::new(PARAMETER_SERVER_EXECUTOR_NAME)
-                    .into_executor(AggregateExecutorConfig {
-                        updates: Receive::peers(worker_ids.to_vec()),
-                        results: Send::peers(worker_ids.to_vec(), SelectionStrategy::All),
-                        optimizer: diloco_config.outer_optimizer.clone(),
-                    })
-                    .into(),
-            },
-            &[parameter_server],
-        )
-        .await
-        .into_diagnostic()?;
+    let parameter_server_id = Arc::new(RwLock::new(None));
 
-        let mut metrics_bridge = match config.status_bridge() {
-            Some(value) => MetricsBridge::new(Box::new(AimConnector::new(value))),
-            None => MetricsBridge::new(Box::new(NoOpConnector::new())),
-        };
+    let mut metrics_bridge = match config.status_bridge() {
+        Some(value) => MetricsBridge::new(Box::new(AimConnector::new(value))),
+        None => MetricsBridge::new(Box::new(NoOpConnector::new())),
+    };
 
-        metrics_bridge.register_stream(ReceiverStream::new(metrics_rx));
+    // Spawn dispatcher for parameter servers.
+    let parameter_dispatcher = {
+        let network = network.clone();
+        let diloco_config = diloco_config.clone();
+        let parameter_server_id = parameter_server_id.clone();
+        let worker_handle = worker_handle.clone();
+
+        tokio::spawn(parameter_pool.for_each_concurrent(None, move |worker| {
```

Comment:
Nit: Let's use `parameter_server` instead of `worker` here, `worker` is a bit overloaded in this context:


```suggestion
        tokio::spawn(parameter_pool.for_each_concurrent(None, move |parameter_server| {
```